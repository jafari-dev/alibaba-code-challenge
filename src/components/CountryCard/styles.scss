.country-card {
  width: 240px;
  border-radius: 8px;
  background-color: #ffffff;
  box-shadow: 0 0 6px #CCCCCC;
  padding-bottom: 16px;

  img {
    width: 100%;
    height: 180px;
    border-top-left-radius: 8px;
    border-top-right-radius: 8px;
  }

  h3, h4 {
    margin-inline: 16px;
  }

  h3 {
    font-weight: 800;
    margin-top: 12px;
    margin-bottom: 16px;
    font-size: 18px;
  }

  h4 {
    font-weight: 600;
    font-size: 14px;
    margin-block: 4px;
    
    span:first-of-type {
      font-weight: bold;
      margin-right: 4px;
    }
  }
}
