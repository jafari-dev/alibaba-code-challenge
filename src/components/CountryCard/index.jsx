import { memo } from "react";
import { Link } from "react-router-dom";
import "./styles.scss";

function CountryCard({name, flagUrl, population, region, capital}) {
  return (
    <div className="country-card">
      <img src={flagUrl} alt={`Flag of ${name}`} />
      <Link to={`/countries/${name.toLowerCase()}`}>
        <h3 data-testid="name">{name}</h3>
      </Link>
      <h4>
        <span>Population:</span>
        <span data-testid="population">{population}</span>
      </h4>
      <h4>
        <span>Region:</span>
        <span data-testid="region">{region}</span>
      </h4>
      <h4>
        <span>Capital:</span>
        <span data-testid="capital">{capital}</span>
      </h4>
    </div>
  )
}

export default memo(CountryCard);
