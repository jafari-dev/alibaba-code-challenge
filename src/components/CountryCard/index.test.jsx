/* eslint-disable testing-library/prefer-screen-queries */
import { render } from "@testing-library/react";
import { faker } from "@faker-js/faker";
import CountryCard from ".";
import { BrowserRouter } from "react-router-dom";

beforeAll(() => {
  faker.seed(1);
});

describe("CountryCard component", () => {
  test("matches the snapshots.", () => {
    const { container } = render(
      <BrowserRouter>
        <CountryCard
          name={faker.address.country()}
          capital={faker.address.city()}
          population={faker.datatype.number({min: 200000, max: 200000000})}
          region={faker.random.word()}
          flagUrl={faker.image.imageUrl()}
        />
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test("sets the props on the suitable element tags", () => {
    const name = faker.address.country();
    const capital = faker.address.city();
    const population = faker.datatype.number({min: 200000, max: 200000000});
    const region = faker.random.word();

    const { getByTestId } = render(
      <BrowserRouter>
        <CountryCard
          name={name}
          capital={capital}
          population={population}
          region={region}
          flagUrl={faker.image.imageUrl()}
        />
      </BrowserRouter>
    );

    expect(getByTestId("name")).toHaveTextContent(name);
    expect(getByTestId("capital")).toHaveTextContent(capital);
    expect(getByTestId("population")).toHaveTextContent(population);
    expect(getByTestId("region")).toHaveTextContent(region);
  });

  test(`sets the 'flagUrl' prop as the source of img tag and sets a suitable \
string as its alt attribute.`, () => {
    const name = faker.address.country();
    const flagUrl = faker.image.imageUrl();

    const { getByRole } = render(
      <BrowserRouter>
        <CountryCard
          name={name}
          flagUrl={flagUrl}
          capital={faker.address.city()}
          population={faker.datatype.number({min: 200000, max: 200000000})}
          region={faker.random.word()}
        />
      </BrowserRouter>
    );

    expect(getByRole("img")).toHaveAttribute("src", flagUrl);
    expect(getByRole("img")).toHaveAttribute("alt", `Flag of ${name}`);
  });
});
