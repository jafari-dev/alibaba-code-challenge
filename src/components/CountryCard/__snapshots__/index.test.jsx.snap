// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryCard component matches the snapshots. 1`] = `
<div>
  <div
    class="country-card"
  >
    <img
      alt="Flag of Iraq"
      src="https://loremflickr.com/640/480"
    />
    <h3
      data-testid="name"
    >
      Iraq
    </h3>
    <h4>
      <span>
        Population:
      </span>
      <span
        data-testid="population"
      >
        186524961
      </span>
    </h4>
    <h4>
      <span>
        Region:
      </span>
      <span
        data-testid="region"
      >
        North
      </span>
    </h4>
    <h4>
      <span>
        Capital:
      </span>
      <span
        data-testid="capital"
      >
        Potomac
      </span>
    </h4>
  </div>
</div>
`;
